File: iosApp/iosApp/Modules/Settings ✅/SettingsWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - SettingsWireframeFactory

protocol SettingsWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: SettingsWireframeContext
    ) -> SettingsWireframe
}

// MARK: - DefaultSettingsWireframeFactory

final class DefaultSettingsWireframeFactory {
    private let settingsService: SettingsService
    private let settingsServiceActionTrigger: SettingsServiceActionTrigger

    init(
        settingsService: SettingsService,
        settingsServiceActionTrigger: SettingsServiceActionTrigger
    ) {
        self.settingsService = settingsService
        self.settingsServiceActionTrigger = settingsServiceActionTrigger
    }
}

extension DefaultSettingsWireframeFactory: SettingsWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: SettingsWireframeContext
    ) -> SettingsWireframe {
        DefaultSettingsWireframe(
            parent,
            context: context,
            settingsService: settingsService,
            settingsServiceActionTrigger: settingsServiceActionTrigger
        )
    }
}

// MARK: - Assembler

final class SettingsWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> SettingsWireframeFactory in
            DefaultSettingsWireframeFactory(
                settingsService: resolver.resolve(),
                settingsServiceActionTrigger: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/settings/SettingsWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.settings

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.settings.SettingsWireframe
import com.sonsofcrypto.web3walletcore.modules.settings.SettingsWireframeContext
import com.sonsofcrypto.web3walletcore.services.settings.SettingsService
import com.sonsofcrypto.web3walletcore.services.settings.SettingsServiceActionTrigger
import smartadapter.internal.extension.name

interface SettingsWireframeFactory {
    fun make(parent: Fragment?, context: SettingsWireframeContext): SettingsWireframe
}

class DefaultSettingsWireframeFactory(
    private val settingsService: SettingsService,
    private val settingsServiceActionTrigger: SettingsServiceActionTrigger,
): SettingsWireframeFactory {

    override fun make(parent: Fragment?, context: SettingsWireframeContext): SettingsWireframe {
        return DefaultSettingsWireframe(
            parent?.let { WeakRef(it) },
            context,
            settingsService,
            settingsServiceActionTrigger,
        )
    }
}

class SettingsWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(SettingsWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultSettingsWireframeFactory(
                it.resolve(SettingsService::class.name),
                it.resolve(SettingsServiceActionTrigger::class.name),
            )
        }
    }
}